kratos/utilities/pointer_communicator.h

```diff
@@ -0,0 +1,454 @@
+    GlobalPointerCommunicator(const DataCommunicator& rComm, GlobalPointersVector< TPointerDataType >& rGpList ):
+        mrDataCommunicator(rComm)
+    {
+        AddPointers(rGpList.begin(),rGpList.end());
+
+        if(mrDataCommunicator.IsDistributed())
+        {
+            ComputeCommunicationPlan();
+        }
+    }
+
+    template< class TIteratorType >
+    GlobalPointerCommunicator(const DataCommunicator& rComm, TIteratorType itBegin, TIteratorType itEnd):
+        mrDataCommunicator(rComm)
+    {
+        AddPointers(itBegin,itEnd);
+
+        if(mrDataCommunicator.IsDistributed())
+        {
+            ComputeCommunicationPlan();
+        }
+    }
+
+    template< class TFunctorType >
+    GlobalPointerCommunicator(DataCommunicator& rComm, TFunctorType&& rFunctor):
+        mrDataCommunicator(rComm)
+    {
+        if(rComm.IsDistributed())
+        {
+            auto gps = rFunctor(rComm);
+            AddPointers(gps.begin(), gps.end());
+            ComputeCommunicationPlan();
+        }
+    }
+
+
+    /// Destructor.
+    virtual ~GlobalPointerCommunicator() {}
+
+    /**this function applies the input function onto the remote data
+     * and retrieves the result to the caller rank
+     */
+    // template< class TSendType >
+    // ResultsProxy<TPointerDataType, TSendType> Apply(std::function< TSendType(GlobalPointer<TPointerDataType>&) > user_function)
+    template< class TFunctorType >
+    ResultsProxy<
+    TPointerDataType,
+    TFunctorType //unfortunately this is deprecated in c++17, so we will have to change this call in the future
+    > Apply(TFunctorType&& UserFunctor)
+    {
+        typedef typename ResultsProxy<TPointerDataType, TFunctorType >::TSendType SendType;
+
+        if(mrDataCommunicator.IsDistributed())
+        {
+            const int current_rank = mrDataCommunicator.Rank();
+
+            GlobalPointersUnorderedMap< TPointerDataType, SendType > non_local_data;
+
+            Update(UserFunctor, non_local_data);
+
+            // //sendrecv data
+            // for(auto color : mColors)
+            // {
+            //     if(color >= 0) //-1 would imply no communication
+            //     {
+            //         auto& gps_to_be_sent = mNonLocalPointers[color];
+
+            //         //TODO: pass Unique to the mNonLocalPointers[recv_rank]
+
+            //         auto recv_global_pointers = SendRecv(gps_to_be_sent, color, color );
+
+            //         std::vector< SendType > locally_gathered_data; //this is local but needs to be sent to the remote node
+            //         for(auto& gp : recv_global_pointers)
+            //             locally_gathered_data.push_back( user_functor(gp) );
+
+            //         auto remote_data = SendRecv(locally_gathered_data, color, color );
+
+            //         for(unsigned int i=0; i<remote_data.size(); ++i)
+            //             non_local_data[gps_to_be_sent[i]] = remote_data[i];
+            //     }
+            // }
```

Comment:
Shouldn't we remove this?

Reply:
please don't, if something goes wrong this is fundamental to debug it, and it takes a while to rewrite it